api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,656 @@
 package aws
 
-import "fmt"
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"sort"
+	"strconv"
+
+	"github.com/zpatrick/go-bytesize"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/api/job"
+	"github.com/quintilesims/layer0/api/provider"
+	awsc "github.com/quintilesims/layer0/common/aws"
+	"github.com/quintilesims/layer0/common/config"
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+var defaultPorts = []int{
+	22,
+	2376,
+	2375,
+	51678,
+	51679,
+}
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]*models.Resource
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+const (
+	Memory = iota
+	CPU
+)
+
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+	}
 }
 
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	// GET RESOURCE PROVIDERS
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceProviders for env '%s': %#v", environmentID, resourceProviders)
+
+	// GET RESOURCE CONSUMERS
+	resourceConsumers, err := e.getResourceConsumers(clusterName)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceConsumers for env '%s': %#v", environmentID, resourceConsumers)
+
+	_, err = e.scale(clusterName, resourceProviders, resourceConsumers)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *EnvironmentScaler) scale(clusterName string, providers []*models.Resource, consumers []*models.Resource) (*models.ScalerRunInfo, error) {
```

Comment:
It would be nice to reduce the overall spaghetti code and/or improve the naming of some of these functions. I still really like the pattern we landed on in the `provider/aws` layer where we used the main exported function (in this case, `Scale`) as a 'master function' for calling all the related helpers. We could do something similar here and I think it would greatly aid clarity.

Reply:
I agree completely.

Reply:
I agree completely.

Reply:
I think you'll find some relief over in #440. :D